Automattic/portkey-android
# Text Editor: Support custom styles

Closes #185. Adds a button to the text editor which cycles through six text styles. The text styles each have a particular typeface, along with rules like text size, line spacing, letter spacing, and shadow.

![font-changing-demo-cropped](https://user-images.githubusercontent.com/9613966/90211008-846dc300-de2a-11ea-880e-c3eaf1cdacaf.gif)

## Remaining TODOs
### General
* [x] AddedView implementation
* [x] Consider serializing to portable units -> Deferred until text feature is done
* [x] ~Look into fallback fonts for other scripts, probably create a new issue~ -> #498, basically `Not-MVP`
* [x] Test in WPAndroid in case of any theming issues

### PR feedback
* [x] Add fallback mechanism when downloadable fonts aren't available

### Design feedback
* [x] Shrink font button and text to align with surrounding image buttons
* [x] Capitalize sentences in text editor
* [x] Reduce scrim overlay by 10%
* [x] ~Adjust scrim overlay further pending feedback~
* [x] ~Adjust line height for Oswald, maybe others~
* [x] ~Support smart quotes/apostrophes~ -> #503, `Not-MVP`
* [x] Decrease line height for Space Mono and Shrikhand
* [x] Increase letterspacing for Shrikhand
* [x] Decrease shadow length for Shrikhand
* [x] Hide text in background while it's being edited
* [x] ~Fix clipping of text on the left for certain fonts and letters~ -> #508, will fix in its own PR
* [x] Add padding to added TextViews so line breaks are kept in the same place as they were in the editor

## Implementation notes

### Font loading

I went with [Android downloadable fonts](https://developer.android.com/guide/topics/ui/look-and-feel/downloadable-fonts), which allows us to save on apk size considerably by letting Android asynchronously download the fonts (or not if they already exist on the device because another app downloaded them already) instead of bundling them in our apk. There is one exception, Oswald, which I added directly as a `ttf` file:

#### Oswald

The designs called for Oswald to be uppercase. There isn't a variation of the font like this, so I looked into a few approaches. A simple solution is to change the input filter to all caps, and turn the text uppercase. We wouldn't want to permanently make all the text uppercase though, since someone just cycling through fonts would have to re-write their text. I started to back up the text and restore it after cycling through Oswald, which works but it's still not clear what should happen when changes are made to the text while Oswald is selected.

I looked at some other apps - some seem to do what I did, but some were clearly not modifying the input filter because the keyboard wasn't forced to all caps. One possible way they might have done this would be to have a custom `EditText` implementation that has separate 'inputted text' and 'displayed text'. This isn't trivial though, and we'd need to keep track of this when handing the text off to be added as a `TextView` in the canvas, and back to being edited.

Instead, I opened Oswald in a font glyph editor and created a modified version where all the lowercase glyphs were replaced with their uppercase versions.

So for example, the glyph for `U+0061`, which is lowercase `a`, is mapped to a glyph that looks like `A`.

This way we can avoid any custom uppercase logic in the code, and all text entered is respected without us having to do anything, because the underlying code points are maintained.

The cost is that we have to include the font file in our apk, which is ~100KB. Since it's the only one that feels okay for now, we could even reduce the size by removing Cyrillic support which is included in Oswald, but is pretty much the only non-Latin Extended script any of our custom fonts support.

### `IdentifiableTypeface`

Stories does some text serializing/deserializing since anything we created in the text editor is replicated in a TextView, and back. This isn't a problem for most text styling (size, letter spacing, color, etc), but is a problem for typefaces. The reason is that once you set a view's `Typeface`, there's no way I could fine to reverse lookup that back into a font ID, so we can't tell what rules go with that typeface, and we can't even serialize it.

I solved this by using a custom TextView which deprecates the usual typeface methods and instead supports a custom `IdentifiableTypeface` class, which is just a wrapper for `Typeface` with a unique ID.

### `TextStyler` and `FontResolver`

We're starting to pass a lot of style parameters back and forth between `PhotoEditor` and `ComposeLoopFrameActivity`. I decided to encapsulate this in a `TextStyler` class that can be passed around more easily. Since `PhotoEditor` probably shouldn't know what custom fonts we support, I'm also using a `FontResolver` interface to let the stories module tell `PhotoEditor` which typeface is associated with which ID.

## To test
1. Add text
2. Push button
3. Repeat 😀

But also,
1. Check that all style attributes are kept when going between the editor and the canvas
2. Turn on Don't Keep Activities and confirm that all style information on the text is preserved
3. Test on earlier APIs
